ggml/src/ggml-vulkan.cpp

```diff
@@ -6441,31 +6464,26 @@ GGML_CALL static ggml_status ggml_backend_vk_graph_compute(ggml_backend_t backen
     // Reserve tensor context space for all nodes
     ctx->tensor_ctxs.resize(cgraph->n_nodes);
 
-    for (int i = 0; i < cgraph->n_nodes; i++) {
-        ggml_vk_build_graph(ctx, cgraph->nodes[i], i, i == last_node, false);
-    }
+    bool first_node_in_batch = true; // true if next node will be first node in a batch
+    int submit_node_idx = 0; // index to first node in a batch
 
+    // submit work every submit_count node to overlap CPU cmdbuffer generation with GPU execution
+    constexpr int submit_count = 50;
     for (int i = 0; i < cgraph->n_nodes; i++) {
-        ggml_tensor * node = cgraph->nodes[i];
-
-        if (ggml_vk_is_empty(node)) {
-            continue;
+        if (first_node_in_batch) {
+            submit_node_idx = i;
         }
 
-        bool ok = ggml_vk_compute_forward(ctx, node, i);
-        if (!ok) {
-            if (node->op == GGML_OP_UNARY) {
-                std::cerr << __func__ << ": error: op not supported UNARY " << node->name << " (" << ggml_unary_op_name(static_cast<ggml_unary_op>(node->op_params[0])) << ")" << std::endl;
-            } else {
-                std::cerr << __func__ << ": error: op not supported " << node->name << " (" << ggml_op_name(node->op) << ")" << std::endl;
-            }
+        // TODO probably it's better to move the submit conter to ggml_vk_build_graph since a lot of nodes might not contain actual vulkan work
```

Comment:
I think we should try this, maybe count only if not `ggml_vk_is_empty(node)`.

Did you try a few values for `submit_count`?

Reply:
I gave a few values a try and smaller values are better. As long as submit + cmdbuffer creation doesn't get more expensive than GPU execution time smaller values will always win since GPU execution starts more early.

I was wondering about counting the return values being true as well. In the end the logic is more complicated now the logic to submit is distributed across two functions now. It'd be better to keep the logic at a single place.

Reply:
I also tested a few values and 100 was very slightly better than 50, but the difference was negligible. Only counting actual submissions also made no difference. So it's probably fine as it is.